src/nplinker/scoring/metcalf_scoring.py

```diff
@@ -76,250 +104,226 @@ def setup(npl):
-        # element Spectra, second MolFams)
-
-        new_results = []
-        met_objs_list = [self.npl.spectra, self.npl.molfams]
-
-        # iterate over the Spectrum results and then the MolFam results
-        for m, met_objs in enumerate(met_objs_list):
-            new_src, new_dst, new_sco = [], [], []
-
-            # go through each pair of input objects and calculate their standardised scores
-            for i in range(len(results[m][self.R_SRC_ID])):
-                gcf = self.npl.gcfs[int(results[m][self.R_SRC_ID][i])]
-                gen_strains = len(gcf.strains)
-
-                # met_obj will now be either a Spectrum or a MolecularFamily, but
-                # doesn't matter which (in this implementation at least) because they
-                # both have a .strains attribute which is the only thing we need. For
-                # Spectra it's the number of strains, for a MolFam it's the total
-                # number of *unique* strains across all Spectra in that family.
-                met_obj = met_objs[int(results[m][self.R_DST_ID][i])]
-                met_strains = len(met_obj.strains)
-
-                # lookup expected + variance values based on strain counts
-                expected = linkfinder.metcalf_expected[met_strains][
-                    gen_strains]
-                variance_sqrt = linkfinder.metcalf_variance_sqrt[met_strains][
-                    gen_strains]
-
-                # calculate the final score based on the basic Metcalf score for these two
-                # particular objects
-                final_score = (results[m][self.R_SCORE][i] -
-                               expected) / variance_sqrt
-
-                # finally apply the scoring cutoff and store the result
-                if self.cutoff is None or (final_score >= self.cutoff):
-                    new_src.append(int(results[m][self.R_SRC_ID][i]))
-                    new_dst.append(int(results[m][self.R_DST_ID][i]))
-                    new_sco.append(final_score)
-
-            # overwrite original "results" with equivalent new data structure
-            new_results.append(np.array([new_src, new_dst, new_sco]))
-
-        return new_results
-
-    def get_links(self, objects, link_collection):
-        # enforce constraint that the list must contain a set of identically typed objects
-        if not all(isinstance(x, type(objects[0])) for x in objects):
-            raise Exception(
-                'MetcalfScoring: uniformly-typed list of objects is required')
-
-        # also can't handle BGCs here, must be one of the other 3 types (GCF/Spectrum/MolecularFamily)
-        if isinstance(objects[0], BGC):
-            raise Exception(
-                'MetcalfScoring requires input type GCF/Spectrum/MolecularFamily, not BGC'
+    def get_links(self, *objects: tuple[GCF, ...] | tuple[Spectrum, ...]
+                  | tuple[MolecularFamily, ...],
+                  link_collection: LinkCollection) -> LinkCollection:
```

Comment:
Why `tuple[Spectrum,...]` ? What is the differenc to using a list or iterable?

Reply:
Good question! The automatically revealed type for `*args` is `tuple`, so I gave it `tuple`. After checking, it's good enough to provide only the value in the tuple, e.g. only  `GCF`.
Details see [PEP 484 [Arbitrary argument lists and default argument values](https://peps.python.org/pep-0484/#arbitrary-argument-lists-and-default-argument-values)](https://peps.python.org/pep-0484/#arbitrary-argument-lists-and-default-argument-values).